### ISSUE TITLE
In-Game: [YYC] Decrementing a struct variable within an if() does not work if it appears after using the "and" keyword in the if()

### ISSUE THREAD

**Issue by SantiFerre** [YoYoGames/GameMaker-Bugs#6337]
### Description

I'm having problems with one line of code when compiling https://github.com/SantiFerre/AlarmCallback with YYC:
`if (_global.__alarmArray[_i].state == __ALARM_STATE_PLAY and --_global.__alarmArray[_i].count <= 0) { //Trigger alarm`

The `_global.__alarmArray[_i].count` variable is never decremented:
![image](https://github.com/YoYoGames/GameMaker-Bugs/assets/15669932/fb3b82cf-4669-4792-9933-20d5df913c13)

If I compile with VM, everything works perfectly.

If I separate the `if` in two, the code works perfectly in YYC, the problem is when nested with an `and` or `&&`
```
if (_global.__alarmArray[_i].state == __ALARM_STATE_PLAY) {
    if (--_global.__alarmArray[_i].count <= 0) { //Trigger alarm
        //CODE
    }
}
```

The error is found in this line of __AlarmSystem.gml:
![image](https://github.com/YoYoGames/GameMaker-Bugs/assets/15669932/57434be7-6187-497e-acd4-4d4ba05447bb)

### Expected Change

The count variable is decremented and loops correctly (as it does on VM)

### Steps To Reproduce

1. Start GameMaker and open [AlarmCallback-main.zip](https://github.com/user-attachments/files/15912558/AlarmCallback-main.zip)
2. Run project attached with YYC
3. Click the AlarmFire button in the debug GUI
4. Observe the count value does not change at all (as shown above)
5. (Optional) Run the game using VM and observe the code works correctly

### How reliably can you recreate this issue using your steps above?

Always

### Which version of GameMaker are you reporting this issue for?

2024.4.1 (Monthly)

### Which platform(s) are you seeing the problem on?

Windows, Android

### Contact Us Package Attached?

- [X] I have attached my Contact Us Package

### Sample Project Added?

- [x] I have included a small sample project

**Comment by YYDan** [comment]
Reproduced using 2024.4.1.202. Added screenshots of the problem and clarified the repro steps and other info above.

Still occurs in IDE v2024.600.0.579  Runtime v2024.600.0.605.

**Comment by YYDan** [comment]
Note that if I split the if into two parts as you say above this does fix the decrement, however I still get different results between VM and YYC. VM gives the same result as when the if was all on the one line, but YYC never loops and instead just keeps decrementing. E.g.:
![image](https://github.com/YoYoGames/GameMaker-Bugs/assets/15669932/48158980-f079-4370-a44a-aaf6c0fdd935)

Worth checking yourself just now, in case there is something else going on, but also for us to be aware of when verifying the fix.

**Comment by SantiFerre** [comment]
Thank you for informing me, I hadn't realised. It is a good solution to correct the problem. I also noticed that in line 81, there is the same problem with the variable decrement and it never decrements the number of repetitions so it is always infinite.
![image](https://github.com/YoYoGames/GameMaker-Bugs/assets/32647282/812cabad-19e5-4448-b2f9-acbe53d0c3c5)

**Comment by YYDan** [comment]
Ahh, yes - of course ;)

**Comment by coyo-t** [comment]
If im reading this right, this isn't a bug, its something called "short circuiting"
basically, for a logical `and` operation, if the left hand side evaluates to `false`, then there is no need to evaluate the right hand side (since the condition will never succeed anyway)
the same goes for an `or` operation; if the left hand side evaluates to `true`, the right hand side isn't evaluated since the condition has already succeeded anyways

in this case, your state check fails, so your decrement on the right side of the `and` never occurs in the first place. the same goes for your later repeat count check and decrement

i dont know if the docs explain what short circuiting is? i havent checked.

**Comment by SantiFerre** [comment]
This is not the case, the `if()` is programmed to do the short-circuit function you mention, but in this case, when the variable is decremented within an if, it does not occur. It has nothing to do with the hierarchy of the `if()` check. 

We have also discovered that if the variable is decremented inside the if, the second check does not do it. Such that `if (--_global.__alarmArray[_i].count <= 0)`. In this case it decrements the variable, but never checks if it is `<= 0` and returns `true`.

**Comment by rwkay** [comment]
Fixed in 2024.8 - in a compound short circuit expression the statements that required a store to a variable was not being handled correctly (the final value to the `--` was not being stored into the global)

**Comment by mgeddesGM** [comment]
verified as of Beta IDE v2024.800.0.593  Runtime v2024.800.0.617
